init.cc

```diff
@@ -41,7 +41,12 @@ std::set<gms::inet_address> get_seeds_from_db_config(const db::config& cfg) {
         }
     }
     if (seeds.empty()) {
-        seeds.emplace(gms::inet_address("127.0.0.1"));
+        if (cfg.developer_mode()) {
+            seeds.emplace(gms::inet_address("127.0.0.1"));
+        } else {
+            startlog.error("Bad configuration: no 'seeds' provided");
+            throw bad_configuration_error();
+        }
```

Comment:
It makes sense, but it needs some mention in the getting started documentation.

Also, once the node has joined, it doesn't really need the seeds. But at this point we don't know this yet.

Reply:
> Also, once the node has joined, it doesn't really need the seeds. But at this point we don't know this yet.

Yes, I mentioned that in https://github.com/scylladb/scylladb/issues/15836#issue-1960960099

I think we can defer that for now.
My initial thought about this are to introduce a new config option, like `seed-nodes-first-boot` (modeled after replace_node_first_boot), that will initialize `seeds` as done here, but the gossiper in this case will use `seeds` only for bootstrap/repair and once done, the gossiper will just use the endpoints loaded from system.peers as seeds for gossip kick-starting purposes.

Reply:
ok, but we do need something in the documentation to avoid suprising people.

Do docker runs default to developer mode? We need to ensure that our getting started documentation isn't invalidated.

Reply:
Of course.
I'll review the user documentation with that regard.
Regarding docker, I'm totally ignorant about it configures seeds.
Where is the code for it?  Who's the owner/expert?

Reply:
dist/docker, @syuu1228

Reply:
Re: docker
https://github.com/scylladb/scylladb/blob/600822379d1adfcc720d5d8470d97924671503cf/dist/docker/scyllasetup.py#L112-L116

But:
https://github.com/scylladb/scylladb/blob/master/docs/dev/docker-hub.md#--seeds-seeds
needs updating.  Will do

Reply:
Sorry for thinking of this only now, but perhaps instead of checking for developer mode, we should check for listen_address/broadcast_listen_address != localhost.

If listen_address was changed, it means we need a seed.

If listen_address was not changed, then a seed won't help. Thus, it's a better discriminant of single-node vs multi-node clusters.


Note listen_address = localhost is legal for docker clusters, but then broadcast_address must be set, so we must check both.

Reply:
Makes sense, though in the long term an already bootstrapped node doesn't need a seed node as it can just use the peer addresses to "seed" the gossip protocol.
This way we can pass the seed as a command line option to new nodes and leave it empty in scylla.yaml

Reply:
Command line options aren't helpful, people run scylladb via systemd.

The way to change this is to start the database in "idle" mode, and issue `nodetool bootstrap contact-point` to join.